src/storage/core.rs

```diff
@@ -518,6 +536,47 @@ impl<K: Key> Storage<K> {
         Ok(blobs)
     }
 
+    fn should_save_corrupted_blob(error: &anyhow::Error) -> bool {
+        if let Some(error) = error.downcast_ref::<Error>() {
+            return match error.kind() {
+                ErrorKind::Bincode(_) | ErrorKind::Validation(_) => true,
+                _ => false,
+            };
+        }
+        false
+    }
+
+    async fn save_corrupted_blob(path: &Path, corrupted_dir_name: &str) -> Result<()> {
+        let parent = path.parent().ok_or_else(|| {
+            anyhow::anyhow!("[{}] blob path don't have parent directory", path.display())
+        })?;
+        let file_name = path
+            .file_name()
+            .ok_or_else(|| anyhow::anyhow!("[{}] blob path don't have file name", path.display()))?
+            .to_os_string();
+        let corrupted_path = parent.join(corrupted_dir_name).join(file_name);

```

Comment:
Here we have lost the directory level with a timestamp, so it will be impossible to properly recover corrupted BLOBs. You have to add the name of the directory inside which the blob is placed: 
`let corrupted_path = parent.join(corrupted_dir_name).join(cur_blob_dir).join(file_name);`

Reply:
It's ok here. `parent` is directory with blobs. `path` argument contains path to blob.